File: android/app/src/main/java/app/notomorrow/feature/progress/WeeklyStatsCard.kt
package app.notomorrow.feature.progress

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.unit.dp
import app.notomorrow.designsystem.Eyebrow
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NTCard
import app.notomorrow.designsystem.NtSegmented
import app.notomorrow.designsystem.SectionHeader
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.WeeklyVolumeChart
import app.notomorrow.feature.workout.workoutCount
import app.notomorrow.feature.workout.workoutSetCount
import app.notomorrow.model.WeightUnit
import app.notomorrow.util.Fmt
import app.notomorrow.util.S
import app.notomorrow.util.rememberNtStrings

/**
 * Progress > Lifts: the last eight weeks of training as small bars, switchable between workouts,
 * volume, time and sets, with this week's value against last week's — port of
 * `WeeklyStatsCard.swift`. [weeks] is [WeeklyStats.weeks], oldest first.
 */
@Composable
fun WeeklyStatsCard(
    weeks: List<WeeklyStats.Week>,
    unit: WeightUnit,
    modifier: Modifier = Modifier,
) {
    var metric by rememberSaveable { mutableStateOf(WeeklyStats.Metric.Workouts) }
    val titles = WeeklyStats.Metric.entries.associateWith { stringResource(it.titleRes) }
    val strings = rememberNtStrings()
    val workoutsLabel = weeks.associateWith { workoutCount(it.workouts) }
    val setsLabel = weeks.associateWith { workoutSetCount(it.sets) }
    val format: (WeeklyStats.Week?) -> String = { week ->
        when {
            week == null -> "—"
            metric == WeeklyStats.Metric.Workouts -> workoutsLabel.getValue(week)
            metric == WeeklyStats.Metric.Volume -> Fmt.volume(week.volumeKg, unit)
            metric == WeeklyStats.Metric.Duration -> Fmt.duration(week.durationSeconds, strings)
            else -> setsLabel.getValue(week)
        }
    }
    val thisWeek = weeks.lastOrNull()
    val lastWeek = if (weeks.size >= 2) weeks[weeks.size - 2] else null
    val ratio = WeeklyStats.change(weeks, metric)
    val chartDescription = titles.getValue(metric) + ": " + weeks.joinToString(", ") { format(it) }

    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(10.dp)) {
        SectionHeader(title = stringResource(S.stats_title))
        NTCard {
            Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
                NtSegmented(
                    options = WeeklyStats.Metric.entries,
                    selected = metric,
                    onSelect = { metric = it },
                    label = { titles.getValue(it) },
                )
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalAlignment = Alignment.Bottom,
                ) {
                    Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                        Eyebrow(stringResource(S.progress_thisWeek))
                        TabularText(format(thisWeek), style = NT.Fonts.title2, color = NT.Colors.ink)
                    }
                    Spacer(Modifier.weight(1f))
                    Column(
                        horizontalAlignment = Alignment.End,
                        verticalArrangement = Arrangement.spacedBy(4.dp),
                    ) {
                        Eyebrow(stringResource(S.workout_history_lastWeek))
                        TabularText(format(lastWeek), style = NT.Fonts.subheadline, color = NT.Colors.ink2)
                    }
                }
                if (ratio != null) {
                    TabularText(
                        text = Fmt.signedPercent(ratio) + " " + stringResource(S.progress_vsLastWeek),
                        style = NT.Fonts.footnote,
                        color = if (ratio >= 0) NT.Colors.ember else NT.Colors.ink2,
                    )
                }
                // The volume chart's bars, fed the selected metric: its axis only labels the weeks.
                WeeklyVolumeChart(
                    weeks = weeks.map { it.toBar(metric) },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(86.dp)
                        .clearAndSetSemantics { contentDescription = chartDescription },
                )
            }
        }
    }
}

File: NoTomorrow/Features/Progress/WeeklyStatsCard.swift
import SwiftUI
import SwiftData

/// Progress > Lifts: the last eight weeks of training as small bars, switchable between workouts, volume, time and
/// sets, with this week's value against last week's. The bucketing lives in `WeeklyStats`.
struct WeeklyStatsCard: View {
    var unit: WeightUnit

    @Query(filter: #Predicate<Workout> { $0.endedAt != nil }, sort: \Workout.startedAt, order: .reverse)
    private var workouts: [Workout]
    @State private var metric: WeeklyStats.Metric = .workouts

    var body: some View {
        let sessions: [WeeklyStats.Session] = workouts.compactMap { workout in
            let sets = workout.completedSetCount
            guard sets > 0 else { return nil }
            return WeeklyStats.Session(startedAt: workout.startedAt, duration: workout.duration,
                                       volumeKg: workout.totalVolumeKg, sets: sets)
        }
        // Monday-first like the rest of the app (calendar, muscles this week).
        var calendar = Calendar.current
        calendar.firstWeekday = 2
        let weeks = WeeklyStats.weeks(sessions: sessions, now: .now, calendar: calendar)
        let thisWeek = weeks.last
        let lastWeek = weeks.count >= 2 ? weeks[weeks.count - 2] : nil
        let ratio = WeeklyStats.change(weeks, metric: metric)

        VStack(alignment: .leading, spacing: 10) {
            SectionHeader(title: "stats.title")
            NTCard {
                VStack(alignment: .leading, spacing: 14) {
                    ProgressSegmented(options: WeeklyStats.Metric.allCases, label: Self.title, selection: $metric)
                    HStack(alignment: .bottom, spacing: 12) {
                        VStack(alignment: .leading, spacing: 4) {
                            Text("progress.thisWeek").eyebrow()
                            Text(verbatim: format(thisWeek))
                                .font(NT.Fonts.title2).foregroundStyle(NT.Colors.ink).tabular()
                                .lineLimit(1).minimumScaleFactor(0.7)
                        }
                        Spacer(minLength: 0)
                        VStack(alignment: .trailing, spacing: 4) {
                            Text("workout.history.lastWeek").eyebrow()
                            Text(verbatim: format(lastWeek))
                                .font(NT.Fonts.subheadline).foregroundStyle(NT.Colors.ink2).tabular()
                                .lineLimit(1).minimumScaleFactor(0.7)
                        }
                    }
                    if let ratio {
                        HStack(spacing: 0) {
                            Text(verbatim: Fmt.signedPercent(ratio) + " ")
                            Text("progress.vsLastWeek")
                        }
                        .font(NT.Fonts.footnote)
                        .foregroundStyle(ratio >= 0 ? NT.Colors.ember : NT.Colors.ink2)
                        .tabular()
                        .lineLimit(1)
                    }
                    // The volume chart's bars, fed the selected metric: its axis only labels the weeks.
                    WeeklyVolumeChart(weeks: weeks.map {
                        WeekVolume(weekStart: $0.weekStart, volumeKg: $0.value(metric), isCurrent: $0.isCurrent)
                    })
                    .frame(height: 86)
                    .accessibilityElement(children: .ignore)
                    .accessibilityLabel(Text(Self.title(metric)))
                    .accessibilityValue(Text(verbatim: weeks.map { format($0) }.joined(separator: ", ")))
                }
            }
        }
    }

    static func title(_ metric: WeeklyStats.Metric) -> LocalizedStringKey {
        switch metric {
        case .workouts: "stats.workouts"
        case .volume: "workout.volume"
        case .duration: "workout.time"
        case .sets: "workout.sets"
        }
    }

    private func format(_ week: WeeklyStats.Week?) -> String {
        guard let week else { return "—" }
        switch metric {
        case .workouts: return WorkoutStrings.workouts(week.workouts)
        case .volume: return Fmt.volume(week.volumeKg, unit: unit)
        case .duration: return Fmt.duration(week.duration)
        case .sets: return WorkoutStrings.sets(week.sets)
        }
    }
}
